platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/files/DataFileCompactor.java

```diff
@@ -0,0 +1,483 @@
+     * there are more / less calls to resume compactions than to pause, or if they are called in a
+     * wrong order, it will result in deadlocks.</b>
+     *
+     * @throws IOException If an I/O error occurs
+     */
+    public void resumeCompaction() throws IOException {
+        try {
+            if (compactionWasInProgress.getAndSet(false)) {
+                assert currentCompactionWriter.get() == null;
+                assert currentCompactionReader.get() == null;
+                startNewCompactionFile(compactionLevelInProgress.getAndSet(0));
+            }
+        } finally {
+            snapshotCompactionLock.release();
+        }
+    }
+
+    /**
+     * Compact (merge) all files that match the given filter.
+     *
+     * @param reportDurationMetricFunction   function to report how long compaction took, in ms
+     * @param reportSavedSpaceMetricFunction function to report how much space was compacted, in Mb
+     * @throws IOException          if there was a problem merging
+     * @throws InterruptedException if the merge thread was interupted
+     * @return true if compaction was performed, false otherwise
+     */
+    @SuppressWarnings("unchecked")
+    public boolean compact(
+            CASableLongIndex index,
+            @Nullable final BiConsumer<Integer, Long> reportDurationMetricFunction,
+            @Nullable final BiConsumer<Integer, Double> reportSavedSpaceMetricFunction)
+            throws IOException, InterruptedException {
+
+        final String storeName = dataFileCollection.getStoreName();
+        final List<? extends DataFileReader<?>> allCompactableFiles = dataFileCollection.getAllCompletedFiles();
+        int allFilesCount = allCompactableFiles.size();
+        if (allFilesCount < getMinNumberOfFilesToCompact()) {
+            logger.debug(
+                    MERKLE_DB.getMarker(),
+                    "[{}] No need to compact as {} is less than the minimum {} files to merge.",
+                    storeName,
+                    allFilesCount,
+                    getMinNumberOfFilesToCompact());
+            return false;
+        }
+
+        final List<DataFileReader<?>> filesToCompact =
+                compactionPlan(getMinNumberOfFilesToCompact()).apply((List<DataFileReader<?>>) allCompactableFiles);
+        if (filesToCompact.isEmpty()) {
+            logger.debug(MERKLE_DB.getMarker(), "[{}] No need to compact, as the compaction plan is empty", storeName);
+            return false;
+        }
+
+        final int filesCount = filesToCompact.size();
+        logger.info(MERKLE_DB.getMarker(), "[{}] Starting compaction", storeName);
+
+        int highestExistingCompactionLevel =
+                filesToCompact.get(filesCount - 1).getMetadata().getCompactionLevel();
+        // target compaction level should  not exceed `maxCompactionLevel` config parameter
+        final int targetCompactionLevel = Math.min(highestExistingCompactionLevel + 1, config.maxCompactionLevel());
```

Comment:
So, every time we get 7 files at level 5 we effectively do full compaction and create a new file at level 5. That file will be ever growing in size. Do we really want it? Why have `maxCompactionLevel` at all?

Reply:
Every time we get 7 files at level 5 + 8 files at level 4 + 8 files at level 3 + ... + 8 files at level 0 we're going to do effectively full compaction. I assume this should happen extremely rarely. We need limit for levels for 2 reasons:
- relatively predictable cadence for full compactions. Even if for the data that rarely changes we need do to it once in a while. If the data gets to the level 6, it's stuck there for long, long time even if it becomes garbage. If I remember our conversation correctly, we decided to give it a shot and see how it behaves in practice. If this level limit proves to be useless, we can always remove it
- we have metrics for each level. I assume we can't have unlimited number of metrics. Therefore, we have this limit. This is a secondary argument but it's something to be taken into account as well.